lensless/unrolled_admm.py

```diff
@@ -0,0 +1,266 @@
+            # overwrite already defined Psi and PsiT
+            self._Psi = psi
+            self._PsiT = psi_adj
+            self._PsiTPsi = psi_gram(self._padded_shape)
+
+        self._PsiTPsi = self._PsiTPsi.to(self._psf.device)
+
+        self.reset()
+
+    def _Psi(self, x):
+        """
+        Operator to map image to space that the image is assumed to be sparse
+        in.
+        """
+        return finite_diff(x)
+
+    def _PsiT(self, U):
+        """
+        Adjoint of `_Psi`.
+        """
+        return finite_diff_adj(U)
+
+    def reset(self, batch_size=1):
+        # needed because ReconstructionAlgorithm initializer call reset to early
+        if not hasattr(self, "_mu1_p"):
+            return
+
+        # ensure that mu1, mu2, mu3, tau are positive
+        self._mu1 = torch.abs(self._mu1_p)
+        self._mu2 = torch.abs(self._mu2_p)
+        self._mu3 = torch.abs(self._mu3_p)
+        self._tau = torch.abs(self._tau_p)
+
+        # TODO initialize without padding
+        if batch_size == 1:
+            self._image_est = torch.zeros(self._padded_shape, dtype=self._dtype).to(
+                self._psf.device
+            )
+        else:
+            self._image_est = torch.zeros((batch_size, *self._padded_shape), dtype=self._dtype).to(
+                self._psf.device
+            )
+        self._X = torch.zeros_like(self._image_est)
+        self._U = torch.zeros_like(self._Psi(self._image_est))
+        self._W = torch.zeros_like(self._X)
+        if self._image_est.max():
+            # if non-zero
+            self._forward_out = self._convolver.convolve(self._image_est)
+            self._Psi_out = self._Psi(self._image_est)
+        else:
+            self._forward_out = torch.zeros_like(self._X)
+            self._Psi_out = torch.zeros_like(self._U)
+
+        self._xi = torch.zeros_like(self._image_est)
+        self._eta = torch.zeros_like(self._U)
+        self._rho = torch.zeros_like(self._X)
+
+        # precompute_R_divmat
+        self._R_divmat = 1.0 / (
+            self._mu1[:, None, None, None]
+            * (torch.abs(self._convolver._Hadj * self._convolver._H))[None, :, :, :]
+            + self._mu2[:, None, None, None] * torch.abs(self._PsiTPsi)[None, :, :, :]
+            + self._mu3[:, None, None, None]
+        ).type(self._complex_dtype)
+
+        # precompute_X_divmat
+        self._X_divmat = 1.0 / (
+            self._convolver._pad(torch.ones_like(self._psf))[None, :, :, :]
+            + self._mu1[:, None, None, None]
+        )
+
+    def _U_update(self, iter):
+        """Total variation update."""
+        # to avoid computing sparse operator twice
+        self._U = soft_thresh(
+            self._Psi_out + self._eta / self._mu2[iter], self._tau[iter] / self._mu2[iter]
+        )
+
+    def _X_update(self, iter):
+        # to avoid computing forward model twice
+        # self._X = self._X_divmat * (self._xi + self._mu1 * self._forward_out + self._data)
```

Comment:
remove comment (and others)

Reply:
A separate pull request should probably be made to do the same with ADMM.